x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/AbstractProfilingPersistenceManager.java

```diff
@@ -0,0 +1,226 @@
+                logger.warn(
+                    "Index [{}] is closed. This is likely to prevent Universal Profiling from functioning correctly",
+                    metadata.getIndex()
+                );
+                return Status.CLOSED;
+            }
+            final IndexRoutingTable routingTable = state.getRoutingTable().index(metadata.getIndex());
+            ClusterHealthStatus indexHealth = new ClusterIndexHealth(metadata, routingTable).getStatus();
+            if (indexHealth == ClusterHealthStatus.RED) {
+                logger.debug(
+                    "Index [{}] health status is RED, any pending mapping upgrades will wait until this changes",
+                    metadata.getIndex()
+                );
+                return Status.UNHEALTHY;
+            }
+            MappingMetadata mapping = metadata.mapping();
+            if (mapping != null) {
+                @SuppressWarnings("unchecked")
+                Map<String, Object> meta = (Map<String, Object>) mapping.sourceAsMap().get("_meta");
+                int currentIndexVersion;
+                int currentTemplateVersion;
+                if (meta == null) {
+                    logger.debug("Missing _meta field in mapping of index [{}], assuming initial version.", metadata.getIndex());
+                    currentIndexVersion = 1;
+                    currentTemplateVersion = 1;
+                } else {
+                    // we are extra defensive and treat any unexpected values as an unhealthy index which we won't touch.
+                    currentIndexVersion = getVersionField(metadata.getIndex(), meta, "index-version");
+                    currentTemplateVersion = getVersionField(metadata.getIndex(), meta, "index-template-version");
+                    if (currentIndexVersion == -1 || currentTemplateVersion == -1) {
+                        return Status.UNHEALTHY;
+                    }
+                }
+                if (index.getVersion() > currentIndexVersion) {
+                    return Status.NEEDS_VERSION_BUMP;
+                } else if (ProfilingIndexTemplateRegistry.INDEX_TEMPLATE_VERSION > currentTemplateVersion) {
+                    // TODO 8.10+: Check if there are any pending migrations. If none are pending we can consider the index up to date.
+                    return Status.NEEDS_MAPPINGS_UPDATE;
+                } else {
+                    return Status.UP_TO_DATE;
+                }
+            } else {
+                logger.warn("No mapping found for existing index [{}]. Index cannot be migrated.", metadata.getIndex());
+                return Status.UNHEALTHY;
+            }
+        } else {
+            return Status.NEEDS_CREATION;
+        }
+    }
+
+    private int getVersionField(Index index, Map<String, Object> meta, String fieldName) {
+        Object value = meta.get(fieldName);
+        if (value instanceof Integer) {
+            return (int) value;
+        } else if (value == null) {
```

Comment:
else if construct isn't needed here

Reply:
It is needed because there are three conditions:

1. The value is a (non-null) integer
2. The value is null
3. The value is of a different (unexpected) type (e.g. `String`)

Reply:
OK, I meant something like
```
    if (value instanceof Integer) {
        return (int) value;
    }
    if (value == null) {
        logger...
        return -1;
    }
    logger...
    return -1;
```

Reply:
Personally I prefer `if` - `else` statements as I find them safer if the code is modified. In that specific case it actually reduces visual noise as now the log statements fit on one line so I've modified it accordingly.